File: Tests/ControllerEndpointTests.swift
import XCTest

@testable import MihomoMeter

final class ControllerEndpointTests: XCTestCase {
  func testNormalizesLoopbackAddressWithoutScheme() throws {
    let endpoint = try ControllerEndpoint(address: "127.0.0.1:9090")

    XCTAssertEqual(endpoint.baseURL.absoluteString, "http://127.0.0.1:9090")
    XCTAssertEqual(
      try endpoint.httpURL(path: "/version").absoluteString,
      "http://127.0.0.1:9090/version"
    )
    XCTAssertEqual(
      try endpoint.webSocketURL(
        path: "/connections",
        queryItems: [URLQueryItem(name: "interval", value: "500")]
      ).absoluteString,
      "ws://127.0.0.1:9090/connections?interval=500"
    )
  }

  func testAcceptsIPv6LoopbackAddress() throws {
    let endpoint = try ControllerEndpoint(address: "http://[::1]:9090")

    XCTAssertEqual(endpoint.baseURL.absoluteString, "http://[::1]:9090")
  }

  func testRejectsNonLoopbackAddress() {
    XCTAssertThrowsError(try ControllerEndpoint(address: "http://192.168.1.2:9090")) {
      error in
      XCTAssertEqual(error as? ControllerEndpointError, .nonLoopbackAddress)
    }
  }

  func testRejectsMissingPortAndExtraPath() {
    XCTAssertThrowsError(try ControllerEndpoint(address: "http://127.0.0.1")) { error in
      XCTAssertEqual(error as? ControllerEndpointError, .missingOrInvalidPort)
    }

    XCTAssertThrowsError(try ControllerEndpoint(address: "http://127.0.0.1:9090/api")) {
      error in
      XCTAssertEqual(error as? ControllerEndpointError, .unsupportedPath)
    }
  }
}

File: platform/windows/MihomoMeter.Windows.Tests/ControllerEndpointTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MihomoMeter.Windows.Core.Domain;

namespace MihomoMeter.Windows.Tests;

[TestClass]
public sealed class ControllerEndpointTests
{
    [TestMethod]
    public void NormalizesLoopbackAddressWithoutScheme()
    {
        var endpoint = new ControllerEndpoint("127.0.0.1:9090");

        Assert.AreEqual("http://127.0.0.1:9090", endpoint.NormalizedAddress);
        Assert.AreEqual(
            "http://127.0.0.1:9090/version",
            endpoint.HttpUri("/version").AbsoluteUri);
        Assert.AreEqual(
            "ws://127.0.0.1:9090/connections?interval=500",
            endpoint.WebSocketUri("/connections", "interval=500").AbsoluteUri);
    }

    [TestMethod]
    public void AcceptsIpv6LoopbackAddress()
    {
        var endpoint = new ControllerEndpoint("http://[::1]:9090");

        Assert.AreEqual("http://[::1]:9090", endpoint.NormalizedAddress);
    }

    [TestMethod]
    public void RejectsUnsupportedAddresses()
    {
        AssertReason("http://192.168.1.2:9090", ControllerEndpointError.NonLoopbackAddress);
        AssertReason("http://127.0.0.1", ControllerEndpointError.MissingOrInvalidPort);
        AssertReason("http://127.0.0.1:9090/api", ControllerEndpointError.UnsupportedPath);
        AssertReason("http://127.0.0.1:9090?", ControllerEndpointError.InvalidAddress);
        AssertReason("http://127.0.0.1:9090#", ControllerEndpointError.InvalidAddress);
        AssertReason("http://user@127.0.0.1:9090", ControllerEndpointError.InvalidAddress);
    }

    private static void AssertReason(string address, ControllerEndpointError reason)
    {
        try
        {
            _ = new ControllerEndpoint(address);
            Assert.Fail("无效地址不应通过验证。");
        }
        catch (ControllerEndpointException exception)
        {
            Assert.AreEqual(reason, exception.Reason);
        }
    }
}
